include/swift/SIL/BasicBlockData.h

```diff
@@ -0,0 +1,189 @@
+///    block list must not be modified between the construction of two
+///    BasicBlockData which are alive at the same time.
+///
+/// We might add a restriction in the future that the block list must not be
+/// modified at all during the lifetime of a BasicBlockData. But this is
+/// currently not possible, because some tools (like StackNesting) split
+/// critical edges while using BasicBlockData.
+template <typename Data> class BasicBlockData {
+  SILFunction *function;
+  std::vector<Data> data;
+  
+  static int getIndex(SILBasicBlock *block) {
+    assert(block->index >= 0 && "Cannot get data for a new block");
+    return block->index;
+  }
+  
+public:
+
+  /// For iterator over the function's blocks, yielding the block and its data.
+  template <typename BlockIterTy, bool IsConst> class IteratorTy {
+    using DataTy = typename std::conditional<IsConst, const Data, Data>::type;
+    using BasicBlockDataTy = typename std::conditional<IsConst,
+      const BasicBlockData<Data>, BasicBlockData<Data>>::type;
+
+    /// The underlying SILBasicBlock iterator.
+    BlockIterTy blockIter;
+
+    /// Reference to the constructing BasicBlockData. Used to retrieve the data.
+    BasicBlockDataTy &blockData;
+
+    friend class BasicBlockData<Data>;
+    
+    IteratorTy(BlockIterTy blockIter, BasicBlockDataTy &blockData) :
+      blockIter(blockIter), blockData(blockData) {}
+
+  public:
+    /// Result of re-referencing the iterator.
+    struct BlockAndData {
+      SILBasicBlock &block;
+      DataTy &data;
+      
+      BlockAndData(SILBasicBlock &block, DataTy &data) :
+        block(block), data(data) {}
+    };
+
+  
+    IteratorTy & operator++() {  // Preincrement
+      blockIter++; return *this;
+    }
+    IteratorTy & operator++(int) {  // Postincrement
+      IteratorTy tmp = *this; ++*this; return tmp;
+    }
+
+    SILBasicBlock &block() const { return *blockIter; }
+    DataTy &data() const { return blockData[&block()]; }
+
+    BlockAndData operator*() const { return BlockAndData(block(), data()); }
+
+    bool operator==(const IteratorTy &rhs) const {
+      return blockIter == rhs.blockIter;
+    }
+    bool operator!=(const IteratorTy &rhs) const {
+      return blockIter != rhs.blockIter;
+    }
+  };
+
+  using iterator = IteratorTy<SILFunction::iterator, false>;
+  using const_iterator = IteratorTy<SILFunction::iterator, true>;
+  using reverse_iterator = IteratorTy<SILFunction::reverse_iterator, false>;
+  using const_reverse_iterator = IteratorTy<SILFunction::reverse_iterator, true>;
+
+  static Data constructDefaultData(SILBasicBlock *) { return Data(); }
+
+  /// Initialize Data for all basic blocks.
+  ///
+  /// The \p init function must return the initial data for a block. If not
+  /// provided, all data is constructed with the Data() default constructor.
+  BasicBlockData(SILFunction *function,
+                 std::function<Data(SILBasicBlock *block)> init =
```

Comment:
We're supposed to use `llvm::function_ref` for functions that aren't called again after the function they are passed to returns.